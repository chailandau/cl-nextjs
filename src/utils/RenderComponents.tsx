import { FC, Fragment } from 'react';

import { Page_PageSections, Project_ProjectSections } from '@/api/graphqlTypes';
import Image from '@/atoms/Image';
import Hero from '@/components/Hero';
import SingleUse from '@/components/SingleUse';
import Socials from '@/components/Socials';
import TextBlock from '@/molecules/TextBlock/TextBlock';

interface Sections {
    /** Sections to render */
    components?: Page_PageSections[] | Project_ProjectSections[];
}

const RenderComponents: FC<Sections> = ({ components }) => {
    if (!components || !components.length) {
        return null;
    }
    const renderComponent = (
        component: Page_PageSections | Project_ProjectSections
    ) => {
        switch (component?.__typename) {
            case 'HeroBlock':
                return <Hero {...component?.hero} />;
            case 'TextBlock':
                return (
                    <TextBlock
                        heading={component?.heading}
                        richText={component?.text}
                    />
                );
            case 'ImageBlock':
                return (
                    <Image
                        src={component?.image?.url || ''}
                        alt={component?.image?.alt || ''}
                        width={component?.image?.width || 1000}
                        height={component?.image?.height || 1000}
                    />
                );
            case 'SocialsBlock':
                return <Socials socials={component?.socials} />;
            case 'SingleUseBlock':
                return <SingleUse component={component?.singleUse} />;
            default:
                return null;
        }
    };

    return (
        <>
            {components?.map((component) => (
                <Fragment key={component?.id}>
                    {renderComponent(component)}
                </Fragment>
            ))}
        </>
    );
};

export default RenderComponents;
