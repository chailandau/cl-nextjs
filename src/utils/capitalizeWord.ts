/**
 * Capitalizes the first letter of a string.
 *
 * @param {string} str - String to be capitalized.
 * @return {string} Capitalized string.
 */
export const capitalizeWord = (str: string): string => {
    if (!str) {
        return str;
    }

    const firstLetter = str.charAt(0).toUpperCase();
    const restOfWord = str.slice(1);

    return firstLetter + restOfWord;
};
