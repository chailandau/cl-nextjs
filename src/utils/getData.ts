import { graphQLClient } from '@/api/graphqlClient';
import { Query } from '@/api/graphqlTypes';

export const getData = async (QUERY: string) =>
    await graphQLClient.request<Query>(QUERY);

export const getPageData = async (QUERY: string, slug: string) =>
    await graphQLClient.request<Query>(QUERY, { slug });
