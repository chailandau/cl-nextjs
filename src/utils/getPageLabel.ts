import { Nav_MenuItems } from '@/api/graphqlTypes';

export const getPageLabel = (menuItem: Nav_MenuItems) => {
    const customLabel =
        menuItem?.overridePageName &&
        menuItem?.customLabel !== null &&
        menuItem.customLabel;

    return menuItem?.page && (customLabel || menuItem?.page?.title);
};
