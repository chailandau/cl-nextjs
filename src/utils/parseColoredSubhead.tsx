import { JSX } from 'react';

import Text, { TextProps } from '@/atoms/Text';

interface ChildSegment {
    text: string;
    bold?: boolean;
}

export interface ColoredSubheadProps {
    children: ChildSegment[];
}

type ColoredSubhead = JSX.Element[] | null | undefined;

/**
 * Parses a colored subhead and returns an array of Text components with appropriate styling.
 *
 * @param coloredSubhead - An array of objects representing the colored subhead.
 * @returns An array of Text components with appropriate styling, or null if the input is falsy.
 */
export const parseColoredSubhead = (
    coloredSubhead: ColoredSubheadProps[]
): ColoredSubhead | null => {
    if (!coloredSubhead) {
        return null;
    }

    const colors = ['aqua', 'white', 'pink'];

    return coloredSubhead.flatMap((segment) => {
        const boldIndexes: number[] = [];

        return segment?.children?.map((child, index) => {
            const isBold = child?.bold;
            let textBgColor;

            if (isBold) {
                boldIndexes.push(index);
                const position = boldIndexes?.indexOf(index);
                textBgColor = colors[position % colors?.length];
            }

            const textProps = {
                as: 'span' as TextProps['as'],
                size: 'md' as TextProps['size'],
                font: 'secondary' as TextProps['font'],
                backgroundColor: textBgColor
            };

            return (
                <Text key={index} {...textProps}>
                    {child?.text}
                </Text>
            );
        });
    });
};
