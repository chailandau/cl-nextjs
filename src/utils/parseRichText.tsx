import { JSX } from 'react';

import styles from '../atoms/Text/Text.module.scss';

import Text from '@/atoms/Text';
import { textSizes } from '@/atoms/Text/Text';

interface ChildSegment {
    text: string;
}

export interface RichTextProps {
    children: ChildSegment[];
}

type RichText = JSX.Element[] | null | undefined;

/**
 * Parses rich text and returns an array of Text components.
 *
 * @param richText - An array of objects representing rich text.
 * @returns An array of Text components, or null if the input is falsy.
 */
export const parseRichText = (
    richText: RichTextProps[],
    size: (typeof textSizes)[number] = 'xs'
): RichText | null => {
    if (!richText) {
        return null;
    }

    const results: JSX.Element[] = [];

    richText.flatMap(
        (segment) =>
            segment?.children?.forEach((child) => {
                const paragraphs = child.text
                    .split('\n')
                    .filter((paragraph) => paragraph.trim() !== '');

                paragraphs.forEach((par, parIndex) => {
                    results.push(
                        <Text
                            className={styles['rich-text']}
                            size={size}
                            key={parIndex}
                        >
                            {par}
                        </Text>
                    );
                });
            })
    );

    return results;
};
