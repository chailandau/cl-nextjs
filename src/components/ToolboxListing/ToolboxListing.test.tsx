import { composeStory } from '@storybook/react';
import { render, screen } from '@testing-library/react';

import Meta, { Default } from './ToolboxListing.stories';

import { testAxeAndSnapshot } from '@/utils/testHelpers';

const ToolboxListing = composeStory(Default, Meta);

describe('Toolbox Listing', () => {
    it('defaults correctly', () => {
        render(<ToolboxListing data={undefined} />);
        expect(screen.queryByRole('section')).not.toBeInTheDocument();
    });
    testAxeAndSnapshot({
        component: <ToolboxListing />
    });
});
