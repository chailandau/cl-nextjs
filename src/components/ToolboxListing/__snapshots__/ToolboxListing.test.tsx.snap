// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toolbox Listing matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="section"
  >
    <section
      class="flex section toolbox-listing"
    >
      <div
        class="flex section-heading"
      >
        <h2
          class="heading md black"
        >
          Toolbox
        </h2>
        <div
          class="icon"
        >
          <div
            class="flex container"
          >
            <img
              alt="Tools icon"
              data-nimg="1"
              decoding="async"
              height="120"
              loading="lazy"
              src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftools.png&w=256&q=75"
              srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftools.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftools.png&w=256&q=75 2x"
              style="color: transparent;"
              width="120"
            />
          </div>
        </div>
      </div>
      <p
        class="text sm primary"
      >
        If I listed every tool I used, this page would be super long. Here’s some I’ve used a bunch, though!
      </p>
      <ul
        class="grid"
      >
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="React icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Freact.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Freact.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Freact.png&w=384&q=75 2x"
                style="color: transparent;"
                width="166"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            React
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="NextJS icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnext.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnext.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnext.png&w=384&q=75 2x"
                style="color: transparent;"
                width="150"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Next.js
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Redux icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fredux.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fredux.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fredux.png&w=384&q=75 2x"
                style="color: transparent;"
                width="152"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Redux
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="GraphQL Icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgraphql.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgraphql.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgraphql.png&w=384&q=75 2x"
                style="color: transparent;"
                width="137"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            GraphQL
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Typescript icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftypescript.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftypescript.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ftypescript.png&w=384&q=75 2x"
                style="color: transparent;"
                width="150"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            TypeScript
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Git icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgit.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgit.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgit.png&w=384&q=75 2x"
                style="color: transparent;"
                width="150"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Git
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Storybook icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstorybook.png&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstorybook.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstorybook.png&w=256&q=75 2x"
                style="color: transparent;"
                width="128"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Storybook
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Contentful icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fcontentful.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fcontentful.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fcontentful.png&w=384&q=75 2x"
                style="color: transparent;"
                width="137"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Contentful
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Jira icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjira.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjira.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjira.png&w=384&q=75 2x"
                style="color: transparent;"
                width="147"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Jira
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Sass icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fsass.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fsass.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fsass.png&w=384&q=75 2x"
                style="color: transparent;"
                width="191"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            SASS / SCSS
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Figma icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffigma.png&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffigma.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffigma.png&w=256&q=75 2x"
                style="color: transparent;"
                width="111"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Figma
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Styled components icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstyled-components.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstyled-components.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fstyled-components.png&w=384&q=75 2x"
                style="color: transparent;"
                width="148"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Styled Components
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Jest icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjest.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjest.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fjest.png&w=384&q=75 2x"
                style="color: transparent;"
                width="133"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Jest
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Github actions icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgithub-actions.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgithub-actions.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fgithub-actions.png&w=384&q=75 2x"
                style="color: transparent;"
                width="150"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Github Actions
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Firebase icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffirebase.png&w=256&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffirebase.png&w=128&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Ffirebase.png&w=256&q=75 2x"
                style="color: transparent;"
                width="117"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Firebase
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="NodeJS Icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnodejs.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnodejs.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fnodejs.png&w=384&q=75 2x"
                style="color: transparent;"
                width="137"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Node.js
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="MongoDB icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fmongodb.png&w=384&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fmongodb.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fmongodb.png&w=384&q=75 2x"
                style="color: transparent;"
                width="150"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            MongoDB
          </p>
        </li>
        <li>
          <div
            class="icon"
          >
            <div
              class="flex container"
            >
              <img
                alt="Docker icon"
                data-nimg="1"
                decoding="async"
                height="150"
                loading="lazy"
                src="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fdocker.png&w=640&q=75"
                srcset="/_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fdocker.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fcl-payload.cr.chailandau.com%2Ficons%2Fdocker.png&w=640&q=75 2x"
                style="color: transparent;"
                width="199"
              />
            </div>
          </div>
          <p
            class="text xs primary"
          >
            Docker
          </p>
        </li>
      </ul>
    </section>
  </div>
</DocumentFragment>
`;
